lib/weather/api_clients/open_weather.ex

```diff
@@ -0,0 +1,49 @@
+defmodule Weather.ApiClients.OpenWeather do
+  @moduledoc """
+  An API client for making requests to an external OpenWeather service.
+  """
+
+  @type handled_response ::
+          {:ok, map()}
+          | {:unauthorized, binary()}
+          | {:error, binary()}
+
+  @doc """
+  Sends an HTTP request to the OpenWeather API
+  to get weather data for the given city.
+  Processes the response by return code.
+  """
+  @spec get_weather_data(city_name :: binary()) :: handled_response
+  def get_weather_data(city_name) do
+    Application.get_env(:weather, __MODULE__)
+    |> fetch_full_url(city_name)
+    |> HTTPoison.get!()
+    |> decode_response_body()
+    |> handle()
+  end
+
+  @spec fetch_full_url(map(), binary()) :: binary()
+  defp fetch_full_url(open_weather_config, city_name) do
+    open_weather_config[:url] <>
+      "&q=#{String.capitalize(city_name)}" <>
+      "&appid=#{open_weather_config[:app_id]}"
+  end
+
+  @spec decode_response_body(response :: HTTPoison.Response.t()) :: map()
+  defp decode_response_body(response),
+    do: Jason.decode!(response.body)
+
+  @spec handle(response_body :: map()) :: handled_response
+  defp handle(response_body) do
+    case response_body["cod"] do
+      200 ->
+        {:ok, response_body}
+
+      401 ->
+        {:unauthorized, "Invalid API key. Please contact to support."}
+
+      _ ->
+        {:error, String.capitalize(response_body["message"])}
+    end
+  end
```

Comment:
Approach inconsistency. Вы вызываете эти функции в одном пайпе, одну за другой. Но первая почему-то бросает исключение, а вторая возвращает ошибку, если что-то пошло не так.

```suggestion
  @spec decode_response_body(response :: HTTPoison.Response.t()) :: map()
  defp decode_response_body(response),
    do: Jason.decode(response.body)

  @spec handle({:ok, map()} | {:error, any()}) :: handled_response()
  defp handle({:error, reason}), do: {:error, reason}
  defp handle({:ok, %{"code" => 200} = response_body}) do
    {:ok, response_body}
  end
  defp handle({:ok, %{"code" => 401} = response_body}) do
    {:unauthorized, "Invalid API key. Please contact to support."}
  end
  defp handle({:ok, %{"message" => message}}) do
    {:error, String.capitalize("message")}
  end
```

Reply:
Спасибо, внес изменения.